diesel/src/migrations/mod.rs

```diff
@@ -238,6 +346,39 @@ fn search_for_migrations_directory(path: &Path) -> Result<PathBuf, MigrationErro
     }
 }
 
+#[doc(hidden)]
+pub trait Migration {
+    fn version(&self) -> &str;
+    fn run(&self, conn: &SimpleConnection) -> Result<(), RunMigrationsError>;
+    fn revert(&self, conn: &SimpleConnection) -> Result<(), RunMigrationsError>;
+}
+
+#[doc(hidden)]
+pub struct ProgrammaticMigration {
```

Comment:
There's no reason that this needs to be in the main crate, we can stick this in the module generated for the users.